src/physics/boundaryConditions/PeriodicBoundary.cpp

```diff
@@ -0,0 +1,230 @@
+                    exit(EXIT_FAILURE);
+                }
+
+                if (insertionIndex < insertedParticles.size()) {
+                    // simply update, if there is already a particle to reuse
+                    insertedParticles.at(insertionIndex)->setX(haloPosition);
+                    insertedParticles.at(insertionIndex)->setV({ 0, 0, 0 });
+                    insertedParticles.at(insertionIndex)->setF({ 0, 0, 0 });
+                    insertedParticles.at(insertionIndex)->setOldF({ 0, 0, 0 });
+                    insertedParticles.at(insertionIndex)->setM(particle.get().getM());
+                    insertedParticles.at(insertionIndex)->setType(particle.get().getType());
+                    insertedParticles.at(insertionIndex)->setActivity(false);
+                } else {
+                    Particle haloParticle(
+                        haloPosition, { 0, 0, 0 }, particle.get().getM(), particle.get().getType());
+                    haloParticle.setActivity(false);
+                    insertedParticles.push_back(std::make_unique<Particle>(haloParticle));
+                }
+
+                // add the particle to the halo cell
+                LGDSim.getGrid()
+                    .cells[haloCellIndex[0]][haloCellIndex[1]][haloCellIndex[2]]
+                    ->addParticle(*insertedParticles.at(insertionIndex++).get());
+            }
+        }
+    }
+
+    // erase all leftover particles, if there are any
+    if (insertionIndex < insertedParticles.size()) {
+        insertedParticles.erase(
+            insertedParticles.begin() + (long)insertionIndex, insertedParticles.end());
+    }
+}
+
+void PeriodicBoundary::postUpdateBoundaryHandling(Simulation& simulation)
+{
+    const LennardJonesDomainSimulation& LGDSim =
+        static_cast<const LennardJonesDomainSimulation&>(simulation);
+
+    // reset old halo references used previously
+    /*
+     * We need to also delete the particles of the opposite side
+     */
+    for (auto pos : allPositions) {
+        auto iter = LGDSim.getGrid().haloCellIterator(pos);
+        auto b = iter.begin();
+        auto e = iter.end();
+        for (auto haloCellIndex : iter) {
+            LGDSim.getGrid()
+                .cells[haloCellIndex[0]][haloCellIndex[1]][haloCellIndex[2]]
+                ->clearParticles();
+        }
+    }
+
+    for (auto boundaryCellIndex : LGDSim.getGrid().boundaryCellIterator(position)) {
+        ParticleRefList& particles =
+            LGDSim.getGrid()
+                .cells[boundaryCellIndex[0]][boundaryCellIndex[1]][boundaryCellIndex[2]]
+                ->getParticles();
+        auto it = particles.begin();
+        auto end = particles.end();
+        while (it != end) {
+            std::array<double, 3> pos = (*it).get().getX();
+
+            CellIndex actualCellIndex = LGDSim.getGrid().getIndexFromPos(pos);
+            CellType actualCellType = LGDSim.getGrid().determineCellType(actualCellIndex);
```

Comment:
You could also just use `auto` for all these types. But I can also see that sometimes it's easier to keep an overview with explicitly stating the type.